clickhouse-admin/src/context.rs

```diff
@@ -3,84 +3,555 @@
+        }
+    }
+}
+
+pub enum GenerateConfigRequest {
+    /// Generates a configuration file for a server or keeper node
+    GenerateConfig {
+        clickward: Clickward,
+        log: Logger,
+        node_settings: NodeSettings,
+        response: oneshot::Sender<Result<GenerateConfigResult, HttpError>>,
+    },
+}
+
+async fn long_running_generate_config_task(
+    incoming: Receiver<GenerateConfigRequest>,
+    generation_tx: watch::Sender<Option<Generation>>,
+) {
+    while let Ok(request) = incoming.recv_async().await {
+        match request {
+            GenerateConfigRequest::GenerateConfig {
+                clickward,
+                log,
+                node_settings,
+                response,
+            } => {
+                let result = generate_config_and_enable_svc(
+                    generation_tx.clone(),
+                    clickward,
+                    node_settings,
+                );
+                if let Err(e) = response.send(result) {
+                    error!(
+                        &log,
+                        "failed to send value from configuration generation to channel: {e:?}"
+                    );
+                };
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum NodeSettings {
+    Replica { settings: ServerConfigurableSettings, fmri: String },
+    Keeper { settings: KeeperConfigurableSettings, fmri: String },
+}
+
+impl NodeSettings {
+    fn fmri(self) -> String {
+        match self {
+            NodeSettings::Replica { fmri, .. } => fmri,
+            NodeSettings::Keeper { fmri, .. } => fmri,
+        }
+    }
+
+    fn generation(&self) -> Generation {
+        match self {
+            NodeSettings::Replica { settings, .. } => settings.generation(),
+            NodeSettings::Keeper { settings, .. } => settings.generation(),
+        }
+    }
+}
+
+pub fn generate_config_and_enable_svc(
+    generation_tx: watch::Sender<Option<Generation>>,
+    clickward: Clickward,
+    node_settings: NodeSettings,
+) -> Result<GenerateConfigResult, HttpError> {
+    let incoming_generation = node_settings.generation();
+    let generation_rx = generation_tx.subscribe();
+    let current_generation = *generation_rx.borrow();
+
+    // If the incoming generation number is lower, then we have a problem.
+    // We should return an error instead of silently skipping the configuration
+    // file generation.
+    if let Some(current) = current_generation {
+        if current > incoming_generation {
+            return Err(HttpError::for_client_error(
+                Some(String::from("Conflict")),
+                StatusCode::CONFLICT,
+                format!(
+                    "current generation '{}' is greater than incoming generation '{}'",
+                    current,
+                    incoming_generation,
+                )
+            ));
+        }
+    };
+
+    let output = match node_settings {
+        NodeSettings::Replica { ref settings, .. } => {
```

Comment:
I'm surprised clippy didn't complain about the use of `ref` being outdated. I'm fine with it, it's just yelled at me about this recently. FWIW, I usually get around this by doing something like `match &node_settings`.

Reply:
maybe clippy just likes me more

Reply:
jk, I changed it 😄